Support for accelerated networking and network security group for scaleset vms

Madhura-CSI opened the pull request:
Fix for https://github.com/ansible/ansible/issues/34032
##### SUMMARY
Current azure_rm_virtualmachine_scaleset module does not have a support for attaching pre-created network security group to virtual machines in scaleset. Also, there is no provision for enabling accelerated networking for VMs. Adding the support through this PR.
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 
##### COMPONENT NAME
azure_rm_virtualmachine_scaleset

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/centos-dev/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```


##### ADDITIONAL INFORMATION
Extra Parameters added to current module:
1. network_security_group (type='str') -> name of network security group.
2. enable_accelerated_networking (type='bool', default='no') -> flag to be set to true, if user wants to enabled accelerated networking for VM interfaces.

Tested Following Use-Case:
Module execution by specifying existing network security group and flag to enable accelerated networking.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
---
- hosts: localhost
  vars:
    vm_image:
      offer: "{{ image_offer | default('CentOS') }}"
      publisher: "{{ image_publisher | default('OpenLogic') }}"
      sku: "{{ image_sku | default('7.4') }}"
      version: "{{ image_version | default('latest') }}"
  tasks:
    - name: Azure VMs | Create Scaleset
      azure_rm_virtualmachine_scaleset:
        resource_group: test-resource_group
        name: testpr-scale-set
        vm_size: Standard_D3_v2
        capacity: 1
        virtual_network_name: test-network
        subnet_name: test-subnet
        admin_username: centos
        ssh_password_enabled: false
        ssh_public_keys:
          - path: "/home/centos/.ssh/authorized_keys"
            key_data: "ssh-rsa AAAAB3NzaC1yc2EAAAADBQAAAAABAYDL/Op8/NENu+nFHS7INjLYoesYxhrVv975UFQWSBF1AYQBxkiJ5W5uQhvUTeZXF1ebR4xnMmRspWf3AWUIdEwXJglfe7p7lr/V4R/TSot0UAvjdXvRXQelIN+AmS2ZsVjcvmqRbubFmJIclo0OTjddHuX3ccYRD1XDUALxOEipcuemWB83Pt+DZnwpJks9dhHJCUCNEhL7McSt1B5dVNfh0Ng83WK0hh58SVx9b0fKD/eKLH00m7Tm9Ni/cisGR3TINs6TH+is56FPFBmFtT6Orx0D+SZUd44pBdfs8FYKsKsS9WoHYcRtEH1XmjpfmP3MLCwJ2VbLy0rD52GcrA6d centos-dev@localhost.localdomain"
        image: "{{ vm_image }}"
        upgrade_policy: Manual
        network_security_group: "test-sg"
        enable_accelerated_networking: yes
      register: vmset
    - debug:
        msg: "{{ vmset }}"
  environment:
    AZURE_SUBSCRIPTION_ID: "{{ azure_subscription_id | default('') }}"
    AZURE_REGION: "{{ azure_region | default('') }}"
    AZURE_AD_USER: "{{ azure_ad_user | default('') }}"
    AZURE_PASSWORD: "{{ azure_password | default('') }}"
  connection: local
```
Result: VM network interfaces are seen attached to network security group 'test-sg'. Also following output shows accelerated networking enabled for VM interface. Output as below.
```
[centos@testpr-scale-set000001 ~]$ lspci
0000:00:00.0 Host bridge: Intel Corporation 440BX/ZX/DX - 82443BX/ZX/DX Host bridge (AGP disabled) (rev 03)
0000:00:07.0 ISA bridge: Intel Corporation 82371AB/EB/MB PIIX4 ISA (rev 01)
0000:00:07.1 IDE interface: Intel Corporation 82371AB/EB/MB PIIX4 IDE (rev 01)
0000:00:07.3 Bridge: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 02)
0000:00:08.0 VGA compatible controller: Microsoft Corporation Hyper-V virtual VGA
_0001:00:02.0 Ethernet controller: Mellanox Technologies MT27500/MT27520 Family [ConnectX-3/ConnectX-3 Pro Virtual Function]_

[centos@testpr-scale-set000001 ~]$ ethtool -S eth0 | grep vf_
     vf_rx_packets: 5578
     vf_rx_bytes: 5396118
     vf_tx_packets: 17941
     vf_tx_bytes: 3257966
     vf_tx_dropped: 0
```
NOTE: These additional features/parameters are not affecting existing scaleset flow. Have tested normal scaleset creation as well. It worked without issues.

Madhura-CSI committed:
Support for accelerated networking and network security group for scaleset vms

Madhura-CSI commented:
ready_for_review

kyliel commented:
@zikalino , could you please have a review?

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -176,6 +176,15 @@
+    network_security_group:
+        description: Name of network security group to be attached to VMs in scaleset.
+        version_added: "2.5"
+        required: false
```
default value is false, can be removed here

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -834,6 +853,13 @@ def vm_size_is_valid(self):
                 return True
         return False
 
+    def get_network_security_group(self, name):
```
can we get the security group from other resource group? In ansible we can handle this by refering it with resource id and adding a `network_security_group_resource_group` field.

Madhura-CSI replied:
Addressed this with latest changes. Please review.

Madhura-CSI commented:
@yuwzho , I have added support to get network security group from other resource group.
Sample playboook call with extra parameter 'network_security_group_resource_group' is as follows.
```
- name: Azure VMs | Create Scaleset
  azure_rm_virtualmachine_scaleset:
    resource_group: test-resource_group
    name: testpr-scale-set
    vm_size: Standard_D3_v2
    capacity: 1
    virtual_network_name: test-network
    subnet_name: test-subnet
    admin_username: centos
    ssh_password_enabled: false
    ssh_public_keys:
      - path: "/home/centos/.ssh/authorized_keys"
        key_data: "ssh-rsa AAAAB3NzaC1yc2EAAAADBQAAAAABAYDL/Op8/NENu+nFHS7INjLYoesYxhrVv975UFQWSBF1AYQBxkiJ5W5uQhvUTeZXF1ebR4xnMmRspWf3AWUIdEwXJglfe7p7lr/V4R/TSot0UAvjdXvRXQelIN+AmS2ZsVjcvmqRbubFmJIclo0OTjddHuX3ccYRD1XDUALxOEipcuemWB83Pt+DZnwpJks9dhHJCUCNEhL7McSt1B5dVNfh0Ng83WK0hh58SVx9b0fKD/eKLH00m7Tm9Ni/cisGR3TINs6TH+is56FPFBmFtT6Orx0D+SZUd44pBdfs8FYKsKsS9WoHYcRtEH1XmjpfmP3MLCwJ2VbLy0rD52GcrA6d centos-dev@localhost.localdomain"
    image: "{{ vm_image }}"
    upgrade_policy: Manual
    network_security_group: "test-sg"
    enable_accelerated_networking: yes
    network_security_group_resource_group: testresourcegroup2
  register: vmset
```
Please review the changes and let me know if anything needs to be modified.

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -595,6 +617,10 @@ def exec_module(self, **kwargs):
 
+                    network_security_group = None
+                    if self.network_security_group:
+                        network_security_group = self.get_network_security_group(self.network_security_group)
```
since `self.network_security_group_resource_group` is not required, I suggest add a logic to make `self.network_security_group_resource_group = self.network_security_group_resource_group or self.resource_group`, then change the `get_network_security_group(self, name)` to `get_network_security_group(self, resource_group, name)`

Madhura-CSI replied:
Hi @yuwzho, as per the suggestion I have modified the code to include resource group as a parameter to get_network_security_group method. Please review and let me know if you would like to have any more improvements.

Madhura-CSI committed:
changes for review comment #2

yuwzho commented:
@Madhura-CSI Thanks for your quick response. The PR looks very good! 
One more question is about the update. The code seems doesn't handle the security_group update. Also,  need to add some integration tests in https://github.com/cambridge-semantics/ansible/blob/devel/test/integration/targets/azure_rm_virtualmachine_scaleset/tasks/main.yml

Madhura-CSI committed:
integration tests for accelerated networking and network security group

Madhura-CSI commented:
@yuwzho, sorry for replying late on this thread. Just to update you on progress with respect to this PR, I am in process of building integration tests around the change set. You can also find few of them with the latest commit. 
For handling VMSS updates, I think we cannot have update around accelerated networking, as it must be enabled at VM creation phase. I am also checking if we can handle security group related updates as well.
Will update here as soon as the code is ready for reviewing.

yuwzho commented:
@mattclay This PR is unstable since the test subscription exceeding quota limits of Core, any advice of it? Should we request increase vCPU quota or modify the test itself?

zikalino commented:
@yuwzho i think we could change capacity from 2 to 1, it's actually irrelevant for the testcases

mattclay commented:
@yuwzho I agree with @zikalino about reducing the capacity.

In general, when creating integration tests for cloud modules we should be using the least amount of resources practical to get the job done as long as that doesn't significantly slow down tests. That means smaller virtual machines, fewer virtual machines, etc.

yuwzho commented:
@mattclay I get your point. So this case, re-run the integration test can help? @Madhura-CSI Could you please try close the PR and re-open it to trigger another build?

Madhura-CSI committed:
reduced capacity, removed redundant test case from original file

Madhura-CSI commented:
@yuwzho, I searched a bit for handling network security group update to existing scaleset. But this seems not possible, due to following reasons:
1. I did try dissociating NSG from network interface of VM in scaleset, but getting 'DisallowedResourceOperation'.

Madhura-CSI commented:
ready_for_review

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -179,6 +179,19 @@
+            - When creating a virtual machine, if a specific network security group from another resource group should be
+              used, use this parameter to specify the resource group to use.
+        version_added: "2.6"
+    network_security_group:
```
We just reached an agreement of how to support the refer resources in other rg. In this scenario, we just accept the `network_security_group` with `type='raw'`, it can be the name or id, also can be dict contains `resource_group` and `name` of a NSG. No need another `network_security_group_resource_group` param. I have proposed a PR to edit this in other module in #37627

https://github.com/VSChina/ansible/blob/26b6addd423b3d879b73177482fa5b0dbe71fdaa/lib/ansible/modules/cloud/azure/azure_rm_subnet.py#L159

https://github.com/VSChina/ansible/blob/26b6addd423b3d879b73177482fa5b0dbe71fdaa/lib/ansible/modules/cloud/azure/azure_rm_subnet.py#L285-L297

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -843,6 +871,13 @@ def vm_size_is_valid(self):
 
+    def get_network_security_group(self, resource_group, name):
+        try:
+            nsg = self.network_client.network_security_groups.get(resource_group, name)
```
Can here just format the resource id, instead of search it online

Madhura-CSI committed:
Addressed review comment, modified integration tests

Madhura-CSI commented:
ready_for_review

yuwzho commented:
Does this `enable_accelerated_networking` and `nsg` handle the update?

yuwzho commented on test/integration/targets/azure_rm_virtualmachine_scaleset/tasks/main.yml:
```diff
@@ -151,62 +150,140 @@
     that: results.changed
 
-- name: Create VMSS (check mode)
+- name: Delete VMSS
```
Does delete need such many parameters?

yuwzho commented on test/integration/targets/azure_rm_virtualmachine_scaleset/tasks/main.yml:
```diff
@@ -151,62 +150,140 @@
+      sku: Stable
+      version: latest
+
+- name: Create VMSS with security group in same resource group, with accelerated networking.
```
When we test create/delete we usually test in  three parts : check_mode, action (create/delete) and idempotent

yuwzho commented on test/integration/targets/azure_rm_virtualmachine_scaleset/tasks/main.yml:
```diff
@@ -151,62 +150,140 @@
+- name: Create VMSS with security group in same resource group, with accelerated networking.
+  azure_rm_virtualmachine_scaleset:
+    resource_group: "{{ resource_group }}"
+    name: testVMSWithAccNetworking
```
Please use a random number for the name like `testVMSS{{ rpfx }}2` to avoid name conflict when parallel running test

yuwzho commented:
Fixes #41218

Madhura-CSI committed:
Addressing review comments

Madhura-CSI commented:
@yuwzho, Greetings.
I have addressed the review comments on integration tests with the latest commit.
To answer your question about handling updates by accelerated networking and nsg, I think it is not possible w.r.t. current azure implementation(As I mentioned in my comment on May 7, please feel free to cross-verify)

Also, integration tests are failing as per latest shippable results. Please have a look and let me know if I need to address any issue.

Thanks!

Madhura-CSI committed:
test commit

yuwzho commented:
shipit

yuwzho commented:
@yungezz

yungezz commented:
shipit

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -183,6 +183,21 @@
+            - Indicates whether user wants to allow accelerated networking for virtual machines in scaleset being created.
+        version_added: "2.7"
+        type: bool
+        default: false
```
default is false by default so no need to have false

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_scaleset.py:
```diff
@@ -388,6 +403,8 @@ def __init__(self):
             virtual_network_resource_group=dict(type='str'),
             virtual_network_name=dict(type='str', aliases=['virtual_network']),
             remove_on_absent=dict(type='list', default=['all']),
+            enable_accelerated_networking=dict(type='bool', default=False),
```
also default false is not needed

Fred-sun commented:
HI @Madhura-CSI what do you think of @zikalino's advice? Could you please help update the PR? Thanks!

yungezz commented:
hi @Madhura-CSI  could you pls look at the review comments? then we can push for next step.

Madhura-CSI committed:
Removing unnecessary defaults

Madhura-CSI commented:
Hi @yungezz , @Fred-sun , @zikalino , I have addressed your reviewed comments.
Ansibot run seems to be unstable due to core limits of test azure account.
Please take a look and let me know if I need to do any more changes.

Fred-sun commented:
@Madhura-CSI Thanks for your contribution, I will try to reproduce, Thanks!

yuwzho commented:
shipit

yuwzho commented:
@zikalino for a further review

yungezz commented:
shipit

gundalow commented:
Thank you for the PR and the reviews. This has been merged into `devel` and will be released in Ansible 2.7.0.
